{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "31831516",
   "metadata": {},
   "source": [
    "# Q1. Explain why we have to use the Exception class while creating a Custom Exception."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa58cf56",
   "metadata": {},
   "source": [
    "Exceptions are special messages that our program uses to tell us when something unexpected or wrong happens. When we create a custom exception, we are making our own special type of message for situations that are specific to our program.\n",
    "\n",
    "we use the Exception class as a base when creating custom exceptions beacause it's like having a blueprint that already knows how messages should look and how to handle them. The Exception class is like a set of rules for making and dealing with these special messages.\n",
    "\n",
    "When we create our custom exception using the Exception class as the base, we are taking advantage of all the helpful things that the Exception class already knows how to do. It's like building on top of something that's already smart about handling messages. This way, we don't have to start from scratch and can focus on making our specific message unique to our program's needs.\n",
    "\n",
    "By inheriting from the Exception class, we can structure our custom exceptions in a way that makes it easier to manage and handle them appropriately."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4ee79f5",
   "metadata": {},
   "source": [
    "# Q2. Write a python program to print Python Exception Hierarchy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dc8e6f56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import inspect\n",
    "\n",
    "def print_exception_hierarchy():\n",
    "    exceptions = BaseException.__subclasses__()\n",
    "    for exception in exceptions:\n",
    "        print(exception.__name__)\n",
    "        subclasses = exception.__subclasses__()\n",
    "        if subclasses:\n",
    "            print_subclasses(subclasses, 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c68b17b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_subclasses(subclasses, indent):\n",
    "    for subclass in subclasses:\n",
    "        print(' ' * indent + subclass.__name__)\n",
    "        nested_subclasses = subclass.__subclasses__()\n",
    "        if nested_subclasses:\n",
    "            print_subclasses(nested_subclasses, indent + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ac5fd65f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exception\n",
      " TypeError\n",
      "  MultipartConversionError\n",
      "  FloatOperation\n",
      " StopAsyncIteration\n",
      " StopIteration\n",
      " ImportError\n",
      "  ModuleNotFoundError\n",
      "  ZipImportError\n",
      " OSError\n",
      "  ConnectionError\n",
      "   BrokenPipeError\n",
      "   ConnectionAbortedError\n",
      "   ConnectionRefusedError\n",
      "   ConnectionResetError\n",
      "    RemoteDisconnected\n",
      "  BlockingIOError\n",
      "  ChildProcessError\n",
      "  FileExistsError\n",
      "  FileNotFoundError\n",
      "  IsADirectoryError\n",
      "  NotADirectoryError\n",
      "  InterruptedError\n",
      "   InterruptedSystemCall\n",
      "  PermissionError\n",
      "  ProcessLookupError\n",
      "  TimeoutError\n",
      "  UnsupportedOperation\n",
      "  herror\n",
      "  gaierror\n",
      "  timeout\n",
      "  Error\n",
      "   SameFileError\n",
      "  SpecialFileError\n",
      "  ExecError\n",
      "  ReadError\n",
      "  SSLError\n",
      "   SSLCertVerificationError\n",
      "   SSLZeroReturnError\n",
      "   SSLWantReadError\n",
      "   SSLWantWriteError\n",
      "   SSLSyscallError\n",
      "   SSLEOFError\n",
      "  URLError\n",
      "   HTTPError\n",
      "   ContentTooShortError\n",
      "  BadGzipFile\n",
      " EOFError\n",
      "  IncompleteReadError\n",
      " RuntimeError\n",
      "  RecursionError\n",
      "  NotImplementedError\n",
      "   ZMQVersionError\n",
      "   StdinNotImplementedError\n",
      "  _DeadlockError\n",
      "  BrokenBarrierError\n",
      "  BrokenExecutor\n",
      "  SendfileNotAvailableError\n",
      "  ExtractionError\n",
      "  VariableError\n",
      " NameError\n",
      "  UnboundLocalError\n",
      " AttributeError\n",
      "  FrozenInstanceError\n",
      " SyntaxError\n",
      "  IndentationError\n",
      "   TabError\n",
      " LookupError\n",
      "  IndexError\n",
      "  KeyError\n",
      "   NoSuchKernel\n",
      "   UnknownBackend\n",
      "  CodecRegistryError\n",
      " ValueError\n",
      "  UnicodeError\n",
      "   UnicodeEncodeError\n",
      "   UnicodeDecodeError\n",
      "   UnicodeTranslateError\n",
      "  UnsupportedOperation\n",
      "  JSONDecodeError\n",
      "  SSLCertVerificationError\n",
      "  Error\n",
      "  ClassNotFound\n",
      "  ClipboardEmpty\n",
      "  MessageDefect\n",
      "   NoBoundaryInMultipartDefect\n",
      "   StartBoundaryNotFoundDefect\n",
      "   CloseBoundaryNotFoundDefect\n",
      "   FirstHeaderLineIsContinuationDefect\n",
      "   MisplacedEnvelopeHeaderDefect\n",
      "   MissingHeaderBodySeparatorDefect\n",
      "   MultipartInvariantViolationDefect\n",
      "   InvalidMultipartContentTransferEncodingDefect\n",
      "   UndecodableBytesDefect\n",
      "   InvalidBase64PaddingDefect\n",
      "   InvalidBase64CharactersDefect\n",
      "   InvalidBase64LengthDefect\n",
      "   HeaderDefect\n",
      "    InvalidHeaderDefect\n",
      "    HeaderMissingRequiredValue\n",
      "    NonPrintableDefect\n",
      "    ObsoleteHeaderDefect\n",
      "    NonASCIILocalPartDefect\n",
      "  IllegalMonthError\n",
      "  IllegalWeekdayError\n",
      "  MacroToEdit\n",
      "  ParserError\n",
      "  InvalidFileException\n",
      "  InvalidVersion\n",
      "  InvalidSpecifier\n",
      "  InvalidMarker\n",
      "  UndefinedComparison\n",
      "  UndefinedEnvironmentName\n",
      "  InvalidRequirement\n",
      "   RequirementParseError\n",
      " AssertionError\n",
      " ArithmeticError\n",
      "  FloatingPointError\n",
      "  OverflowError\n",
      "  ZeroDivisionError\n",
      "   DivisionByZero\n",
      "   DivisionUndefined\n",
      "  DecimalException\n",
      "   Clamped\n",
      "   Rounded\n",
      "    Underflow\n",
      "    Overflow\n",
      "   Inexact\n",
      "    Underflow\n",
      "    Overflow\n",
      "   Subnormal\n",
      "    Underflow\n",
      "   DivisionByZero\n",
      "   FloatOperation\n",
      "   InvalidOperation\n",
      "    ConversionSyntax\n",
      "    DivisionImpossible\n",
      "    DivisionUndefined\n",
      "    InvalidContext\n",
      " SystemError\n",
      "  CodecRegistryError\n",
      " ReferenceError\n",
      " MemoryError\n",
      " BufferError\n",
      " Warning\n",
      "  UserWarning\n",
      "   GetPassWarning\n",
      "   FormatterWarning\n",
      "  DeprecationWarning\n",
      "   ProvisionalWarning\n",
      "  PendingDeprecationWarning\n",
      "  SyntaxWarning\n",
      "  RuntimeWarning\n",
      "   ProactorSelectorThreadWarning\n",
      "   UnknownTimezoneWarning\n",
      "   PEP440Warning\n",
      "  FutureWarning\n",
      "   ProvisionalCompleterWarning\n",
      "  ImportWarning\n",
      "  UnicodeWarning\n",
      "  BytesWarning\n",
      "  ResourceWarning\n",
      "  DeprecatedTzFormatWarning\n",
      "  PkgResourcesDeprecationWarning\n",
      " Error\n",
      " _OptionError\n",
      " error\n",
      " Verbose\n",
      " _Error\n",
      " SubprocessError\n",
      "  CalledProcessError\n",
      "  TimeoutExpired\n",
      " _GiveupOnSendfile\n",
      " error\n",
      " LZMAError\n",
      " RegistryError\n",
      " _GiveupOnFastCopy\n",
      " ZMQBaseError\n",
      "  ZMQError\n",
      "   ContextTerminated\n",
      "   Again\n",
      "   InterruptedSystemCall\n",
      "  ZMQBindError\n",
      "  NotDone\n",
      " Error\n",
      " error\n",
      " PickleError\n",
      "  PicklingError\n",
      "  UnpicklingError\n",
      " _Stop\n",
      " TokenError\n",
      " StopTokenizing\n",
      " ClassFoundException\n",
      " EndOfBlock\n",
      " TraitError\n",
      " ArgumentError\n",
      " ArgumentTypeError\n",
      " ConfigError\n",
      "  ConfigLoaderError\n",
      "   ArgumentError\n",
      "  ConfigFileNotFound\n",
      " ConfigurableError\n",
      "  MultipleInstanceError\n",
      " ApplicationError\n",
      " NoIPAddresses\n",
      " Error\n",
      "  CancelledError\n",
      "  TimeoutError\n",
      "  InvalidStateError\n",
      " Incomplete\n",
      " TimeoutError\n",
      " InvalidStateError\n",
      " LimitOverrunError\n",
      " QueueEmpty\n",
      " QueueFull\n",
      " InvalidPortNumber\n",
      " Empty\n",
      " Full\n",
      " DuplicateKernelError\n",
      " TimeoutError\n",
      " COMError\n",
      " ArgumentError\n",
      " ReturnValueIgnoredError\n",
      " KeyReuseError\n",
      " UnknownKeyError\n",
      " LeakedCallbackError\n",
      " BadYieldError\n",
      " ReturnValueIgnoredError\n",
      " Return\n",
      " ErrorDuringImport\n",
      " BdbQuit\n",
      " Restart\n",
      " FindCmdError\n",
      " HomeDirError\n",
      " ProfileDirError\n",
      " IPythonCoreError\n",
      "  TryNext\n",
      "  UsageError\n",
      "  StdinNotImplementedError\n",
      " InputRejected\n",
      " GetoptError\n",
      " OptionError\n",
      " ErrorToken\n",
      " PrefilterError\n",
      " AliasError\n",
      "  InvalidAliasError\n",
      " Error\n",
      "  InterfaceError\n",
      "  DatabaseError\n",
      "   InternalError\n",
      "   OperationalError\n",
      "   ProgrammingError\n",
      "   IntegrityError\n",
      "   DataError\n",
      "   NotSupportedError\n",
      " Warning\n",
      " SpaceInInput\n",
      " DOMException\n",
      "  IndexSizeErr\n",
      "  DomstringSizeErr\n",
      "  HierarchyRequestErr\n",
      "  WrongDocumentErr\n",
      "  InvalidCharacterErr\n",
      "  NoDataAllowedErr\n",
      "  NoModificationAllowedErr\n",
      "  NotFoundErr\n",
      "  NotSupportedErr\n",
      "  InuseAttributeErr\n",
      "  InvalidStateErr\n",
      "  SyntaxErr\n",
      "  InvalidModificationErr\n",
      "  NamespaceErr\n",
      "  InvalidAccessErr\n",
      "  ValidationErr\n",
      " ValidationError\n",
      " EditReadOnlyBuffer\n",
      " _Retry\n",
      " InvalidLayoutError\n",
      " HeightIsUnknownError\n",
      " ParserSyntaxError\n",
      " InternalParseError\n",
      " _PositionUpdatingFinished\n",
      " SimpleGetItemNotFound\n",
      " UncaughtAttributeError\n",
      " ParamIssue\n",
      " _JediError\n",
      "  InternalError\n",
      "  WrongVersion\n",
      "  RefactoringError\n",
      " OnErrorLeaf\n",
      " InvalidPythonEnvironment\n",
      " MessageError\n",
      "  MessageParseError\n",
      "   HeaderParseError\n",
      "   BoundaryError\n",
      "  MultipartConversionError\n",
      "  CharsetError\n",
      " Error\n",
      " HTTPException\n",
      "  NotConnected\n",
      "  InvalidURL\n",
      "  UnknownProtocol\n",
      "  UnknownTransferEncoding\n",
      "  UnimplementedFileMode\n",
      "  IncompleteRead\n",
      "  ImproperConnectionState\n",
      "   CannotSendRequest\n",
      "   CannotSendHeader\n",
      "   ResponseNotReady\n",
      "  BadStatusLine\n",
      "   RemoteDisconnected\n",
      "  LineTooLong\n",
      " InteractivelyDefined\n",
      " KillEmbedded\n",
      " QueueEmpty\n",
      " QueueFull\n",
      " ExpatError\n",
      " Error\n",
      "  ProtocolError\n",
      "  ResponseError\n",
      "  Fault\n",
      " BadZipFile\n",
      " LargeZipFile\n",
      " error\n",
      " com_error\n",
      " internal_error\n",
      " ParseBaseException\n",
      "  ParseException\n",
      "  ParseFatalException\n",
      "   ParseSyntaxException\n",
      " RecursiveGrammarException\n",
      " ResolutionError\n",
      "  VersionConflict\n",
      "   ContextualVersionConflict\n",
      "  DistributionNotFound\n",
      "  UnknownExtra\n",
      " UnableToResolveVariableException\n",
      " InvalidTypeInArgsException\n",
      "GeneratorExit\n",
      "SystemExit\n",
      "KeyboardInterrupt\n",
      "CancelledError\n",
      "DebuggerInitializationError\n"
     ]
    }
   ],
   "source": [
    "print_exception_hierarchy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d15e094b",
   "metadata": {},
   "source": [
    "# Q3. What errors are defined in the ArithmeticError class? Explain any two with an example."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2c9cbf4",
   "metadata": {},
   "source": [
    "\n",
    "The ArithmeticError class in Python is a base class for exceptions that are related to arithmetic operations. It serves as a superclass for various specific arithmetic-related exception classes. Here are some of the errors that are defined as subclasses of the ArithmeticError class:\n",
    "\n",
    "1.OverflowError: Raised when an arithmetic operation exceeds the limit of representable values, resulting in an overflow.\n",
    "\n",
    "2.ZeroDivisionError: Raised when division or modulo operation is performed with zero as the divisor.\n",
    "\n",
    "3.FloatingPointError: Raised when a floating-point arithmetic operation fails, such as an invalid operation involving NaN (Not-a-Number) or infinity.\n",
    "\n",
    "4.FloatingPointError: Raised when a floating-point arithmetic operation fails, such as an invalid operation involving NaN (Not-a-Number) or infinity.\n",
    "\n",
    "5.ArithmeticError: The base class itself can be raised in certain situations where an arithmetic error occurs, and a more specific exception class doesn't apply"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4b024904",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ZeroDivisionError occurred: Cannot divide by zero\n"
     ]
    }
   ],
   "source": [
    "#ZeroDivisionError\n",
    "\n",
    "numerator = 10\n",
    "denominator = 0\n",
    "\n",
    "try:\n",
    "    result = numerator / denominator  # Raises ZeroDivisionError\n",
    "except ZeroDivisionError:\n",
    "    print(\"ZeroDivisionError occurred: Cannot divide by zero\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e9527a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#OverflowError: \n",
    "\n",
    "try:\n",
    "    result = 10 ** 1000 \n",
    "except OverflowError:\n",
    "    print(\"OverflowError occurred: Result too large to store\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c879ae74",
   "metadata": {},
   "source": [
    "# Q4. Why LookupError class is used? Explain with an example KeyError and IndexError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7b2fc31",
   "metadata": {},
   "source": [
    "The LookupError class in Python is used as the base class for exceptions that are raised when a key or index used on a mapping or sequence is invalid. It provides a way to handle errors related to accessing elements in a collection, such as dictionaries or lists."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4a8d717",
   "metadata": {},
   "source": [
    "***keyerror***\n",
    "\n",
    "The KeyError is raised when you try to access a dictionary key that does not exist."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ed8333ac",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KeyError occurred: Key 'c' not found in dictionary\n"
     ]
    }
   ],
   "source": [
    "my_dict = {'a': 1, 'b': 2}\n",
    "\n",
    "try:\n",
    "    value = my_dict['c']  \n",
    "except KeyError:\n",
    "    print(\"KeyError occurred: Key 'c' not found in dictionary\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5df7383d",
   "metadata": {},
   "source": [
    "***IndexError:***\n",
    "    \n",
    "The IndexError is raised when you try to access an index that is out of range in a sequence, such as a list or a string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "752fbec9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IndexError occurred: Index out of range\n"
     ]
    }
   ],
   "source": [
    "my_list = [10, 20, 30]\n",
    "\n",
    "try:\n",
    "    element = my_list[5]  # Raises IndexError\n",
    "except IndexError:\n",
    "    print(\"IndexError occurred: Index out of range\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0c7e7b6",
   "metadata": {},
   "source": [
    "# Q5. Explain ImportError. What is ModuleNotFoundError?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763eef54",
   "metadata": {},
   "source": [
    "The ImportError is a built-in exception in Python that is raised when an import statement fails to locate and load a module or a component from a module. This exception occurs when the interpreter encounters difficulties while attempting to import a module or its attributes. Importing modules is a fundamental aspect of Python programming, and the ImportError helps handle various scenarios where imports go wrong.\n",
    "\n",
    "common situations in ImportError:-\n",
    "\n",
    "1.Module Not Found\n",
    "2.Circular Imports\n",
    "3.Missing Dependencies\n",
    "4.Incorrect Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1fb5a36b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ModuleNotFoundError\n",
    "#suppose we have module named my_module.py and it contain following functions\n",
    "\n",
    "\n",
    "def greet(name):\n",
    "    print(f\"Hello, {name}!\")\n",
    "\n",
    "def farewell(name):\n",
    "    print(f\"Goodbye, {name}!\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0ccd3c04",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ImportError occurred: No module named 'mymodule'\n"
     ]
    }
   ],
   "source": [
    "#Now,if we have another script that tries to import \n",
    "#functions from my_module, but if we misspell the module name:\n",
    "#then error message will be occured.\n",
    "\n",
    "try:\n",
    "    from mymodule import greet, farewell  # Misspelled module name\n",
    "except ImportError as e:\n",
    "    print(\"ImportError occurred:\", e)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7c3fed",
   "metadata": {},
   "source": [
    "# Q6. List down some best practices for exception handling in python."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f5d0562",
   "metadata": {},
   "source": [
    "1.Always use specific exception\n",
    "\n",
    "2.Print always proper message\n",
    "\n",
    "3.Always try to log error\n",
    "\n",
    "4.Always avoid to write multiple exception handling\n",
    "\n",
    "5.Document all error\n",
    "\n",
    "6.cleanup all resources"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4774d4f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
